source/encoding.c

```diff
@@ -0,0 +1,316 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#include <aws/common/encoding.h>
+
+static const char *hex_chars = "0123456789abcdef";
+
+static const uint8_t base64_sentinal_value = 0xff;
+static const char base64_encoding_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
+static uint8_t base64_decoding_table[256];
+static int8_t base64_decoding_table_init = 0;
+
+static const uint8_t bytes_7 = 7 * 8;
+static const uint8_t bytes_6 = 6 * 8;
+static const uint8_t bytes_5 = 5 * 8;
+static const uint8_t bytes_4 = 4 * 8;
+static const uint8_t bytes_3 = 3 * 8;
+static const uint8_t bytes_2 = 2 * 8;
+static const uint8_t bytes_1 = 1 * 8;
+static const uint8_t mask = 0xff;
+
+
+int aws_hex_encode(const uint8_t *to_encode, size_t to_encode_len, char *output, size_t *output_size) {
+
+    size_t encoded_len = (to_encode_len << 1) + 1;
+
+    if(encoded_len <= to_encode_len) {
+        return aws_raise_error(AWS_ERROR_OVERFLOW_DETECTED);
+    }
+
+    if (!output) {
+        *output_size = encoded_len;
+        return AWS_OP_SUCCESS;
+    }
+
+    if (*output_size < encoded_len) {
+        return aws_raise_error(AWS_ERROR_INVALID_BUFFER_SIZE);
+    }
+
+    size_t written = 0;
+    for (size_t i = 0; i < to_encode_len; ++i) {
+
+        output[written++] = hex_chars[to_encode[i] >> 4 & 0x0f];
+        output[written++] = hex_chars[to_encode[i] & 0x0f];
+    }
+
+    output[written++] = '\0';
+    *output_size = written;
+
+    return AWS_OP_SUCCESS;
+}
+
+static uint8_t char_to_int(char character) {
```

Comment:
make it clear with this function name that this is specifically for hex decoding